Document SDS for ingress gateways

banks opened the pull request:
This PR adds documentation for features added in #10903 and #11163.

These are "low-level" features designed to be used by systems integration builders rather than typical-end users but are included to provide a complete example so this functionality can be used. I don't anticipate a Learn guide for this feature as it's more of an integration interface so included a somewhat guide-like complete example of getting the whole feature to work.

If anyone can think of a better place to put the more detailed guide than the ingress gateway overview page I'd be open to suggestions.

Preview: https://consul-o89qrzjyl-hashicorp.vercel.app/docs/connect/config-entries/ingress-gateway#sds

karl-cardenas-coding commented on website/content/docs/connect/config-entries/service-router.mdx:
```diff
@@ -574,6 +574,69 @@ spec:
+      yaml: false,
+      name: 'Set',
+      type: 'map<string|string>: optional',
+      description: `The set of header value keyed by header name to
+      add. If one or more header values with the same (case-insensitive) name already exist,
+      the value set here will replace them all.
+      If Envoy is used as the proxy, the value may contain
```
```suggestion
      description: `The set of header names to add. If one or more header name values with the same (case-insensitive) name already exist, then the value set here will replace them all.
      If Envoy is used as the proxy, the value may contain
```

@banks does this suggestion still convey what you were aiming for? The original sentence was a bit confusing to read.

karl-cardenas-coding commented on website/content/docs/connect/config-entries/service-router.mdx:
```diff
@@ -574,6 +574,69 @@ spec:
+      yaml: false,
+      name: 'Add',
+      type: 'map<string|string>: optional',
+      description: `The set of header value keyed by header name to
+      add. If a header with the same (case-insensitive) name already
+      exists, the value set here will be appended and both will be present.
+      If Envoy is used as the proxy, the value may contain
```
```suggestion
      description: `The set of header names to add. If one or more header name values with the same (case-insensitive) name already exist, then the value set here will replace them all.
      If Envoy is used as the proxy, the value may contain
```
Same as below, does this make sense?

karl-cardenas-coding reviewed (commented):
Thanks for adding the docs content @banks . I left some suggestions and questions for you. I also tagged @trujillo-adam to review from a tech writing perspective.

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -810,11 +810,29 @@ spec:
+          name: 'SDS',
+          yaml: false,
+          type: 'SDSConfig: <optional>',
+          description: "Configures the gateway to load TLS certificates from an external SDS service. See [SDS](/docs/connect/gateways/ingress-gateway#sds) for more details on usage.<br><br>SDS properties set here will be used as defaults for all listeners on the gateway.",
```
```suggestion
          description: "Defines a set of parameters that configures the gateway to load TLS certificates from an external SDS service. See [SDS](/docs/connect/gateways/ingress-gateway#sds) for more details on usage.<br><br>SDS properties defined in this field are used as defaults for all listeners on the gateway.",
```
I think our docs could use more info about the type of values each config takes as well as their default values.

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -810,11 +810,29 @@ spec:
+            {
+              name: 'ClusterName',
+              type: 'string',
+              description: "The SDS cluster name to connect to to retrieve certificates. This cluster must be [specified in the Gateway's bootstrap configuration](/docs/connect/gateways/ingress-gateway#sds).",
```
```suggestion
              description: "Specifies the name of the SDS cluster from which Consul should retrieve certificates. This cluster must be [specified in the Gateway's bootstrap configuration](/docs/connect/gateways/ingress-gateway#sds).",
```

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -810,11 +810,29 @@ spec:
+            {
+              name: 'CertResource',
+              type: 'string',
+              description: "The SDS resource name to request when fetching the certificate from the SDS service. Setting this causes all listeners to be served exclusively over TLS with this certificate unless overridden by listener-specific TLS configuration.",
```
```suggestion
              description: "Specifies an SDS resource name. Consul will request the SDS resource name when fetching the certificate from the SDS service. Setting this causes all listeners to be served exclusively over TLS with this certificate unless overridden by listener-specific TLS configuration.",
```

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -896,6 +915,69 @@ spec:
+            {
+              name: 'SDS',
+              type: 'SDSConfig: <optional>',
+              description: `Configures the SDS source for the certificate for this specific service.
```
Similar suggestion as line 822

banks replied:
```suggestion
              description: `Defines a set of parameters that configures the SDS source for the certificate for this specific service.
```

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -896,6 +915,69 @@ spec:
+              type: 'bool: false',
+              description: {
+                hcl:
+                  "Set this configuration to enable built-in TLS for this listener.<br><br>If TLS is enabled, then each host defined in each service's `Hosts` field will be added as a DNSSAN to the gateway's x509 certificate. Note that even hosts from other listeners with TLS disabled will be added.",
```
```suggestion
                  "Set this configuration to `true` to enable built-in TLS for this listener.<br><br>If TLS is enabled, then each host defined in each service's `Hosts` field will be added as a DNSSAN to the gateway's x509 certificate. Note that even hosts from other listeners with TLS disabled will be added.",
```

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -896,6 +915,69 @@ spec:
+                hcl:
+                  "Set this configuration to enable built-in TLS for this listener.<br><br>If TLS is enabled, then each host defined in each service's `Hosts` field will be added as a DNSSAN to the gateway's x509 certificate. Note that even hosts from other listeners with TLS disabled will be added.",
+                yaml:
+                  "Set this configuration to enable built-in TLS for this listener.<br><br>If TLS is enabled, then each host defined in the `hosta` field will be added as a DNSSAN to the gateway's x509 certificate. Note that even hosts from other listeners with TLS disabled will be added.",
```
```suggestion
                  "Set this configuration to `true` to enable built-in TLS for this listener.<br><br>If TLS is enabled, then each host defined in the `hosts` field will be added as a DNSSAN to the gateway's x509 certificate. Note that even hosts from other listeners with TLS disabled will be added.",
```

trujillo-adam commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -896,6 +915,69 @@ spec:
+            {
+              name: 'SDS',
+              type: 'SDSConfig: <optional>',
+              description: "Configures the listener to load TLS certificates from an external SDS service. See [SDS](/docs/connect/gateways/ ingress-gateway#sds) for more details on usage.<br><br>SDS properties set here will be used as defaults for all services on this listener.",
```
```suggestion
              description: "Defines a set of parameters that configures the listener to load TLS certificates from an external SDS service. See [SDS](/docs/connect/gateways/ingress-gateway#sds) for more details on usage.<br><br>SDS properties set here will be used as defaults for all services on this listener.",
```

trujillo-adam reviewed (commented):
I had a few questions and made some formatting suggestions that you can take or leave.

banks commented on website/content/docs/connect/config-entries/ingress-gateway.mdx:
```diff
@@ -896,6 +915,69 @@ spec:
+              type: 'SDSConfig: <optional>',
+              description: `Configures the SDS source for the certificate for this specific service.
+                  At least one custom host must be specified in \`Hosts\`.
+                  This certificate will be served to all requests identifying one of the
```
```suggestion
                  The certificate retrieved from SDS will be served for all requests identifying one of the
```

banks commented on website/content/docs/connect/gateways/ingress-gateway.mdx:
```diff
@@ -68,4 +68,212 @@ must also provide `agent:read` for its node's name in order to discover the agen
+~> **Advanced Topic** This is a low-level feature designed for developers
+building integrations with custom TLS management solutions.
+
+Consul 1.11 added support for Ingress Gateways to serve TLS certificates to
```
They are kind of a proper noun in that they are specific components in our product. I followed the lead of the rest of the docs which for the most part seem to capitalize but I don't have a strong opinion either way.

banks commented on website/content/docs/connect/gateways/ingress-gateway.mdx:
```diff
@@ -68,4 +68,212 @@ must also provide `agent:read` for its node's name in order to discover the agen
+
+### 1. Configure Static SDS Cluster(s).
+
+Each Ingress Gateway Envoy instance must have one or more additional [static
+clusters](/docs/connect/proxies/envoy#envoy_extra_static_clusters_json) defined
+when registering the ingress gateway to specify how to connect to the external
+service. This requires a manual registration of the Ingress Gateway proxy rather
```
Heh no it's far from obvious - I was trying to be as precise as possible but that's clearly not been successful or at least not made it easier to understand 😄 .

Basically: an ingress gateway config is "logical" with a name, but then one or more actual Envoy instances are run to implement the gateway. I want to be specific that it's when running those Envoy's that make up this specific gateway that you need to define these extra things. It's not all Envoy instances because Gateways that are not using SDS don't need this, nor do Sidecar proxies in the mesh.

So "Each gateway instance" wasn't precise enough because it's only ingress gateways. "Each Ingress Gateway instance" is confusing because that could be refering to the logical gateway (i.e. it's ConfigEntry). We could say "Each Ingress Gateway Proxy", I avoided that just because I expect this feature to only with with Envoy and it's technically possible that someone could use a different proxy as an ingress (though very unlikely and not something we are expecting to work on any time soon).

How about this as an improvement:
```suggestion
Each Envoy proxy that makes up this Ingress Gateway must define one or more additional [static
clusters](/docs/connect/proxies/envoy#envoy_extra_static_clusters_json) when registering. These additional clusters define how Envoy should connect to the required SDS service(s). Defining extra clusters in Envoy's bootstrap configuration requires a manual registration of the Ingress Gateway with Consul proxy rather
```

trujillo-adam replied:
Yes, I think that's much better

banks commented on website/content/docs/connect/gateways/ingress-gateway.mdx:
```diff
@@ -68,4 +68,212 @@ must also provide `agent:read` for its node's name in order to discover the agen
+clusters](/docs/connect/proxies/envoy#envoy_extra_static_clusters_json) defined
+when registering the ingress gateway to specify how to connect to the external
+service. This requires a manual registration of the Ingress Gateway proxy rather
+than relying on `-register` flag to `consul connect envoy -gateway=ingress`.
```
```suggestion
It's not possible to use the `-register` flag with `consul connect envoy -gateway=ingress` to automatically register the proxy in this case.
```

trujillo-adam commented on website/content/docs/connect/gateways/ingress-gateway.mdx:
```diff
@@ -68,4 +68,208 @@ must also provide `agent:read` for its node's name in order to discover the agen
+The cluster(s) must provide connection information and any necessary
+authentication information such as mTLS credentials.
+
+In this example we will show:
+ - A DNS name to discover the SDS service addresses
+ - Local certificate files for TLS client authentication with the SDS server
+   (the certificates are assumed to be created and managed by some other
+   process)
```
```suggestion
The following example will demonstrate how to use:
 - A DNS name to discover the SDS service addresses.
 - Local certificate files for TLS client authentication with the SDS server. The certificates are assumed to be created and managed by some other  process.
```
Our style guide discourages using the royal "we" (dude) unless guidance on a specific topic is coming from HashiCorp.

trujillo-adam reviewed (approved):
Left a few suggestions.

banks commented:
Thanks for all the reviews folks. This fell off my radar a bit - I had some other work locally that didn't get pushed up, but I'll get the feedback all incorporated this week so it's done and merged in time for GA.

banks commented:
Thanks @trujillo-adam and @freddygv, I finally got around to completing my edits here. I think this addresses all the feedback so far and looks good to me.

If you want to take another pass, feel free otherwise i'll merge this in a few days to make sure it makes it into 1.11 on time!

hc-github-team-consul-core commented:
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/527813.

hc-github-team-consul-core commented:
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/527833.

hc-github-team-consul-core commented:
:cherries::white_check_mark: Cherry pick of commit 131897bff6326ffa3b400be7f7518bfef0d44eaf onto `release/1.11.x` succeeded!
